Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp

```diff
@@ -275,10 +283,24 @@ void ThreadedCompositor::renderLayerTree()
     m_scene->applyStateChanges(states);
     m_scene->paintToCurrentGLContext(viewportTransform, FloatRect { FloatPoint { }, viewportSize }, m_flipY);
 
-    m_context->swapBuffers();
+    if (m_eglSupportsDamage && !m_scene->lastDamagedRects().isEmpty()) {
+        // The damage rect originally has its coordinates origin in the top left corner, as used by CSS, Wayland, etc.
+        // We need to translate to a bottom-left origin, as expected by Mesa/EGL, which already re-translates the y coordinate
+        // internally back to top-left origin for Wayland before sending the damage rectangle.
+
+        // Work on a copy to only pass rects bottom-left as coordinate origin to EGL
+        Vector damagedRects { m_scene->lastDamagedRects() };
+        for (auto& damagedRect : damagedRects) {
+            damagedRect.setY(viewportSize.height() - damagedRect.y() - damagedRect.height());
+        }
+        m_context->swapBuffersWithDamage(damagedRects);
+    } else {
+        m_context->swapBuffers();
+    }
+
```

Comment:
We were not doing any check before because `swapBuffers()` just returns early for surfaceless contexts, but now we are doing more things we don't need. So, I think we should check here if context is surfaceless and simply don't call swap buffers at all.

Reply:
done